dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java

```diff
@@ -107,6 +124,19 @@
         .put( "attributeOptionCombo", "psi_aoc" ).put( "completedDate", "psi_completeddate" )
         .put( "deleted", "psi_deleted" ).build();
 
+    // -------------------------------------------------------------------------
+    // Caching
+    // -------------------------------------------------------------------------
+    private final static int DATA_ELEMENT_TO_CODE_CACHE_THRESHOLD = 200;
+    private final static int DATA_ELEMENT_TO_ATTRIBUTE_CACHE_THRESHOLD = 500;
+    private final static Duration CACHE_DATA_DURATION = Duration.ofMinutes( 15 );
+
+    private Instant dataElementToCodeCacheLastCleared = Instant.now();
+    private Instant dataElementToAttributeCacheLastCleared = Instant.now();
+
+    private CachingMap<String, DataElement> dataElementToCodeCache = new CachingMap<>();
```

Comment:
The cache should not be at the class level due to concurrency issues (multiple threads). It is sufficient to use these caches within the method, i.e. per request.

Reply:
Well. we could use a concurrent version of HashMap which should be safe enough. I wanted to make it more efficient by not fetching the same data over and over. But OK, will move it into the method